### DISCUSSION TITLE
Shifting articles between locations and keep track of them grouped by location

### DISCUSSION THREAD

**Discussion by IT-VBFK** [inventree/InvenTree#3692] [Q&A] [answered]
Hi everyone

I have a quick question (surely I am missing something):

### Status quo
The status quo is that I have one instance of InvenTree which has three stock locations (A, B, C) defined.

### Goal
I have to shift articles from one to an other location, but have to keep track of that items which are shifted (especially the price and the amount), because we have to send a fictitious invoice (mostly for auditing reasons) for this items.

### Example
Let's say in location A we have 

- item1 (qty 5)
- item2 (qty 5)

and we want to shift 3 items each to location B.

As I understand this procedure so far, I have to 

- open a sales order
- select the needed parts (going through the whole process of new order -> shipping and so on)

_And_

- open a purchase order
- go through the whole process of new order -> receive items and so on

### Question
Is there a simpler workaround using the "Stock / Transfer Stock" and keep track of these items

I think about something like this:

- Go to Stock / location A / Transfer stock
- Select item1 (3 pieces) and item2 (3 pieces) and location B and click "submit"
- Go to an overview (which does not exist as far than I can see 🙂) which lists
    - From location A: item1 (3) to location B (3) which value is 3xX €
    - From location A: item2 (3) to location B (3) which value is 3xY €
- Click export for selected items per location

**Comment by SchrodingersGat** [comment, marked as answer]
Each stock "transaction" (e.g. transfer between locations) is tracked as part of the [stock history](https://inventree.readthedocs.io/en/latest/stock/stock/#stock-tracking) so what you are looking for should already be achievable "out of the box".

In the "stock transfer" dialog you can select what quantity to transfer. The stock item is split into the appropriate quantities.

![image](https://user-images.githubusercontent.com/10080325/191206627-fcd68ffd-e396-4b87-bad7-f64c8d6de351.png)

**Reply by IT-VBFK** [reply]
Yeah, I know this "transfer" feature, but the stock history does not contain the needed data, but mostly (in our case) irrelevant data..

Take this for example
<img width="1485" alt="image" src="https://user-images.githubusercontent.com/49762557/191977315-2ee63b55-8c52-4379-8731-63b13d1b87a6.png">

I do not see:

- the "from" location or the "to" location (I do not see if the location presented is the "from" or the "to"
- the calculated value (price) for which we have to write an invoice
- the count of pieces shifted to the other location
- the second row... I do not even know anything (neither the "from" location nor the "destination" location nor the value nor the real stock item count)

What I was expecting was something like this:

| Description | From       | To         | Value  | Pieces |
|-------------|------------|------------|--------|--------|
| Take item 1 | Location A | Location B | 100€   | 5      |
| description for Item 2      | Location B | Location C | 19,99€ | 2      |
| Item 3      | Location A | Location C | 1123€  | 1      |

So, is there a feature or anything else to achieve that without going through the whole process of sell/purchase?

**Reply by matmair** [reply]
@IT-VBFK no. This is not an accounting software.

**Reply by miggland** [reply]
1. the "from" location or the "to" location (I do not see if the location presented is the "from" or the "to"
2. the calculated value (price) for which we have to write an invoice
3. the count of pieces shifted to the other location
4. the second row... I do not even know anything (neither the "from" location nor the "destination" location nor the value nor the real stock item count)

Some of this information is available via API. It requires two steps, since the quantity isn't available directly in the tracking (which I would have assumed it would be...). The example uses the demo server, but the data might be reset, so check the IDs.

The relevant stock item (created by the move) has ID 995 in this example.
https://demo.inventree.org/api/stock/track/?item=995
Gives something like this (only showing the top entry, ie the last move):
```
[
  {
    "pk": 889,
    "item": 995,
    "date": "2022-09-28 10:49",
    "deltas": {
      "location": 11,
      "stockitem": 408,
      "location_detail": {
        "pk": 11,
        "url": "/stock/location/11/",
        "name": "Loose Parts",
        "level": 1,
        "description": "Loose parts / cut tapes",
        "parent": 7,
        "pathstring": "Electronics Lab/Loose Parts",
        "owner": null,
        "icon": ""
      },
      "stockitem_detail": {
        "batch": "2022-4-29",
        "belongs_to": null,
        "build": null,
        "customer": null,
        "delete_on_deplete": true,
        "expiry_date": null,
        "is_building": false,
        "link": "",
        "location": 3,
        "notes": null,
        "owner": null,
        "packaging": "",
        "part": 82,
        "purchase_order": null,
        "pk": 408,
        "quantity": 128,
        "sales_order": null,
        "serial": null,
        "status": 10,
        "status_text": "OK",
        "stocktake_date": null,
        "supplier_part": 204,
        "barcode_hash": "",
        "updated": "2022-09-28",
        "purchase_price": null,
        "purchase_price_currency": "USD",
        "purchase_price_string": "-"
      }
    },
    "label": "Split from parent item",
    "notes": "Test",
    "tracking_type": 40,
    "user": 1
  },
...
]
```

It's not one line, but the data is here that shows:

1. Stock moved from location 3 to location 11 (deltas.stockitem_detail.location -> deltas.location)
2. Purchase price (none given in the example)
3. Date of change: 2022-09-28 10:49

To get the quantity (the stockitem_detail is what remains on the original stock item, not how many parts were moved), check the original stock item. The relevant ID is 408, which you can see above as deltas.stockitem. Check the history for this part:
https://demo.inventree.org/api/stock/track/?item=408
This output contains more than what you are looking for, but using the date of the change you can identify the relevant entry:
```
  {
    "pk": 890,
    "item": 408,
    "date": "2022-09-28 10:49",
    "deltas": {
      "removed": 14,
      "quantity": 131
    },
    "label": "Stock manually removed",
    "notes": "Test",
    "tracking_type": 12,
    "user": 1
  },
```
This gives you the quantity: 14.

Depending on how useful the purchase price is for your invoice, this might be enough. Otherwise, you could maybe use stock metadata (See the documentation at https://demo.inventree.org/api-doc/#stock-metadata-read) for this purpose.

You can either use the API, or create a plugin which reacts on the event? (https://inventree.readthedocs.io/en/latest/extend/plugins/event/#events)... I've never used that though, so you'll have to ask the developers about that.

**Reply by miggland** [reply]
Maybe you could put in a feature request to get the moved quantity to appear in the deltas section in the first API request - imho it does belong there.

**Reply by matmair** [reply]
@IT-VBFK do you want to further discuss this or can we mark this as answered?

**Reply by IT-VBFK** [reply]
Oh.. sorry missed that completely..

sure it's answered
